File: js_example/src/did_resolver.js
import bs58 from 'bs58';

function toBytes(value) {
  if (value == null) return new Uint8Array();
  if (value instanceof Uint8Array) return value;
  if (value instanceof ArrayBuffer) return new Uint8Array(value);
  if (typeof value === 'string') {
    if (value.startsWith('0x')) {
      const hex = value.slice(2);
      const out = new Uint8Array(hex.length / 2);
      for (let i = 0; i < out.length; i += 1) {
        out[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
      }
      return out;
    }
    return new TextEncoder().encode(value);
  }
  if (Array.isArray(value)) {
    return Uint8Array.from(value);
  }
  return new Uint8Array();
}

function toString(value) {
  return new TextDecoder().decode(toBytes(value));
}

function decodeStoredBytes(value) {
  if (typeof value !== 'string' || !value.startsWith('0x')) {
    return value;
  }
  const bytes = toBytes(value);
  const text = toString(bytes);
  const printableAscii = /^[\x20-\x7e]*$/.test(text);
  return printableAscii && (text.startsWith('u') || text.startsWith('{')) ? text : bytes;
}

function parseJsonBytes(value) {
  try {
    return JSON.parse(toString(value));
  } catch {
    return toString(value);
  }
}

function encodeUvarint(value) {
  const out = [];
  let v = Number(value) >>> 0;
  while (true) {
    let b = v & 0x7f;
    v >>>= 7;
    if (v !== 0) b |= 0x80;
    out.push(b);
    if (v === 0) break;
  }
  return Uint8Array.from(out);
}

function decodeUvarint(bytes) {
  let value = 0;
  let shift = 0;
  for (let i = 0; i < bytes.length; i += 1) {
    const byte = bytes[i];
    value |= (byte & 0x7f) << shift;
    if ((byte & 0x80) === 0) {
      return { value, offset: i + 1 };
    }
    shift += 7;
  }
  throw new Error('Invalid uvarint');
}

function toMultikey(publicKey, codec) {
  if (codec == null) return null;
  const raw = toBytes(publicKey);
  const prefix = encodeUvarint(codec);
  const out = new Uint8Array(prefix.length + raw.length);
  out.set(prefix, 0);
  out.set(raw, prefix.length);
  return `u${Buffer.from(out).toString('base64url')}`;
}

export function decodeMultikey(value) {
  const decodedValue = decodeStoredBytes(value);
  const text = typeof decodedValue === 'string' ? decodedValue : toString(decodedValue);
  if (!text.startsWith('u')) {
    return { multicodec: null, publicKey: toBytes(decodedValue) };
  }
  const bytes = Uint8Array.from(Buffer.from(text.slice(1), 'base64url'));
  const { value: multicodec, offset } = decodeUvarint(bytes);
  return { multicodec, publicKey: bytes.slice(offset) };
}

function didIdBytes(did) {
  const prefix = 'did:qsb:';
  if (!did.startsWith(prefix)) {
    throw new Error('DID must start with did:qsb:');
  }
  const didId = bs58.decode(did.slice(prefix.length));
  if (didId.length !== 32) {
    throw new Error('DID id must decode to 32 bytes');
  }
  return didId;
}

export function didToDocument(did, details) {
  const roleMap = {
    Authentication: 'authentication',
    AssertionMethod: 'assertionMethod',
    KeyAgreement: 'keyAgreement',
    CapabilityInvocation: 'capabilityInvocation',
    CapabilityDelegation: 'capabilityDelegation',
  };

  const services = (details.services || []).map((service) => ({
    id: toString(service.id ?? service.service_id),
    type: toString(service.service_type ?? service.serviceType ?? service.type),
    serviceEndpoint: toString(service.endpoint ?? service.serviceEndpoint),
  }));

  const metadata = (details.metadata || []).map((item) => ({
    key: toString(item.key),
    value: toString(item.value),
  }));

  const doc = {
    '@context': ['https://www.w3.org/ns/did/v1', 'https://w3id.org/security/multikey/v1'],
    id: did,
    verificationMethod: [],
    authentication: [],
    assertionMethod: [],
    keyAgreement: [],
    capabilityInvocation: [],
    capabilityDelegation: [],
    service: services,
    metadata,
  };

  (details.keys || [])
    .filter((key) => !(key.revoked ?? false))
    .forEach((key) => {
      const keyId = toString(key.key_id ?? key.keyId);
      const controller = toString(key.controller ?? did);
      const material = normalizeKeyMaterial(key);
      const roles = key.roles || [];
      const vm = {
        id: keyId,
        controller,
      };
      if (material.type === 'Jwk') {
        vm.type = 'JsonWebKey2020';
        vm.publicKeyJwk = parseJsonBytes(material.publicKeyJwk);
      } else {
        const publicKeyMultibase =
          toMultikey(material.publicKey, material.multicodec) ??
          `z${bs58.encode(material.publicKey)}`;
        vm.type = 'Multikey';
        vm.publicKeyMultibase = publicKeyMultibase;
      }
      doc.verificationMethod.push(vm);
      roles.forEach((role) => {
        const field = roleMap[role];
        if (field) doc[field].push(keyId);
      });
    });

  return doc;
}

function normalizeKeyMaterial(key) {
  const material = key.key_material ?? key.keyMaterial;
  const multikey =
    material?.Multikey ??
    material?.multikey ??
    material?.multikeyValue ??
    material?.multikey_value ??
    null;
  if (multikey) {
    const raw = Array.isArray(multikey)
      ? multikey[1]
      : typeof multikey === 'string'
        ? multikey
        : multikey.public_key ?? multikey.publicKey ?? multikey;
    const decoded = decodeMultikey(raw);
    const multicodec = Array.isArray(multikey)
      ? Number(multikey[0])
      : multikey.multicodec ?? multikey.multicodecValue ?? decoded.multicodec ?? null;
    return {
      type: 'Multikey',
      multicodec,
      publicKey: decoded.publicKey,
    };
  }

  const jwk =
    material?.Jwk ??
    material?.jwk ??
    material?.jwkValue ??
    material?.jwk_value ??
    null;
  if (jwk) {
    const raw =
      typeof jwk === 'string'
        ? jwk
        : jwk.public_key_jwk ?? jwk.publicKeyJwk ?? jwk;
    return {
      type: 'Jwk',
      publicKeyJwk: toBytes(decodeStoredBytes(raw)),
    };
  }

  return {
    type: 'Multikey',
    multicodec: key.multicodec ?? null,
    publicKey: toBytes(key.public_key ?? key.publicKey),
  };
}

export async function resolveDid(api, did) {
  const success = (details, warning = null) => ({
    didDocument: didToDocument(did, details),
    didDocumentMetadata: {
      deactivated: details?.deactivated ?? false,
      versionId: details?.version ?? 0,
    },
    didResolutionMetadata: {
      contentType: 'application/did+ld+json',
      error: null,
      ...(warning ? { warning } : {}),
    },
  });
  const notFound = () => ({
    didDocument: null,
    didDocumentMetadata: {
      deactivated: false,
      versionId: 0,
    },
    didResolutionMetadata: {
      contentType: null,
      error: 'notFound',
    },
  });

  try {
    if (api.rpc?.did?.getByString) {
      const opt = await api.rpc.did.getByString(did);
      const result = opt?.toJSON ? opt.toJSON() : null;
      if (result) return success(result);
    }
    const provider = getProvider(api);
    if (provider?.send) {
      const response = await provider.send('did_getByString', [did]);
      const result = response?.result ?? response;
      if (result) {
        if (typeof result === 'string' && result.startsWith('0x')) {
          const decoded = decodeDidDetails(api, result);
          if (decoded) return success(decoded);
        }
        if (typeof result === 'object') {
          if (result.error) {
            const fallback = await queryDidRecords(api, did);
            if (fallback) {
              return success(
                fallback,
                'did_getByString returned an error; resolved through Did.DidRecords storage query'
              );
            }
          }
          return success(result);
        }
      }
    }
  } catch (error) {
    const result = await queryDidRecords(api, did);
    if (result) {
      return success(
        result,
        'did_getByString failed; resolved through Did.DidRecords storage query'
      );
    }
    return {
      didDocument: null,
      didDocumentMetadata: {
        deactivated: false,
        versionId: 0,
      },
      didResolutionMetadata: {
        contentType: null,
        error: 'runtimeApiDecodeError',
        message: error.message,
      },
    };
  }
  return notFound();
}

async function queryDidRecords(api, did) {
  const query = api.query?.did?.didRecords;
  if (!query) return null;
  const record = await query(didIdBytes(did));
  if (record?.isNone) return null;
  const value = record?.unwrap ? record.unwrap() : record;
  return value?.toJSON ? value.toJSON() : null;
}

function decodeDidDetails(api, hex) {
  const candidates = ['Option<DidDetails>', 'Option<did::DidDetails>'];
  for (const typeName of candidates) {
    try {
      const decoded = api.registry.createType(typeName, hex);
      const json = decoded.toJSON?.();
      if (json) return json;
    } catch {
      // try next candidate
    }
  }
  return null;
}

function getProvider(api) {
  return (
    api?.rpc?.provider ||
    api?.rpcCore?.provider ||
    api?._rpcCore?.provider ||
    null
  );
}

File: python_example/src/app/did_resolver.py
import base58
import base64
import json
from substrateinterface.exceptions import SubstrateRequestException


def encode_uvarint(value: int) -> bytes:
    out = bytearray()
    v = value
    while True:
        b = v & 0x7F
        v >>= 7
        if v:
            b |= 0x80
        out.append(b)
        if not v:
            break
    return bytes(out)


def to_multikey(public_key: bytes, codec: int | None) -> str:
    if codec is None:
        return "z" + base58.b58encode(public_key).decode("ascii")
    prefixed = encode_uvarint(codec) + public_key
    encoded = base64.urlsafe_b64encode(prefixed).decode("ascii").rstrip("=")
    return f"u{encoded}"


def _bytes(value) -> bytes:
    if value is None:
        return b""
    if isinstance(value, bytes):
        return value
    if isinstance(value, str):
        if value.startswith("0x"):
            return bytes.fromhex(value[2:])
        return value.encode("utf-8")
    return bytes(value)


def _decode_text(value) -> str:
    return _bytes(value).decode("utf-8", errors="replace")


def _parse_json_bytes(value):
    text = _decode_text(value)
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        return text


def _normalize_key_material(key: dict) -> dict:
    material = key.get("key_material") or key.get("keyMaterial") or {}
    multikey = material.get("Multikey") or material.get("multikey")
    if multikey:
        return {
            "type": "Multikey",
            "multicodec": multikey.get("multicodec"),
            "public_key": _bytes(multikey.get("public_key") or multikey.get("publicKey")),
        }

    jwk = material.get("Jwk") or material.get("jwk")
    if jwk:
        return {
            "type": "Jwk",
            "public_key_jwk": _bytes(jwk.get("public_key_jwk") or jwk.get("publicKeyJwk")),
        }

    return {
        "type": "Multikey",
        "multicodec": key.get("multicodec"),
        "public_key": _bytes(key.get("public_key") or key.get("publicKey")),
    }


def _did_id_bytes(did: str) -> bytes:
    prefix = "did:qsb:"
    if not did.startswith(prefix):
        raise ValueError("DID must start with did:qsb:")
    did_id = base58.b58decode(did[len(prefix) :])
    if len(did_id) != 32:
        raise ValueError("DID id must decode to 32 bytes")
    return did_id


def _success(did: str, result: dict) -> dict:
    return {
        "didDocument": did_to_document(did, result),
        "didDocumentMetadata": {
            "deactivated": result.get("deactivated", False),
            "versionId": result.get("version", 0),
        },
        "didResolutionMetadata": {
            "contentType": "application/did+ld+json",
            "error": None,
        },
    }


def _not_found() -> dict:
    return {
        "didDocument": None,
        "didDocumentMetadata": {
            "deactivated": False,
            "versionId": 0,
        },
        "didResolutionMetadata": {
            "contentType": None,
            "error": "notFound",
        },
    }


def _query_did_records(substrate, did: str) -> dict | None:
    record = substrate.query("Did", "DidRecords", [_did_id_bytes(did)])
    return getattr(record, "value", None)


def did_to_document(did: str, details: dict) -> dict:
    role_map = {
        "Authentication": "authentication",
        "AssertionMethod": "assertionMethod",
        "KeyAgreement": "keyAgreement",
        "CapabilityInvocation": "capabilityInvocation",
        "CapabilityDelegation": "capabilityDelegation",
    }
    services = []
    for service in details.get("services", []):
        service_type = service.get("service_type", service.get("type", b""))
        service_endpoint = service.get("endpoint", service.get("serviceEndpoint", b""))
        services.append(
            {
                "id": bytes(service["id"]).decode("utf-8", errors="replace"),
                "type": bytes(service_type).decode("utf-8", errors="replace"),
                "serviceEndpoint": bytes(service_endpoint).decode("utf-8", errors="replace"),
            }
        )

    metadata = []
    for item in details.get("metadata", []):
        metadata.append(
            {
                "key": bytes(item["key"]).decode("utf-8", errors="replace"),
                "value": bytes(item["value"]).decode("utf-8", errors="replace"),
            }
        )

    doc = {
        "@context": ["https://www.w3.org/ns/did/v1", "https://w3id.org/security/multikey/v1"],
        "id": did,
        "verificationMethod": [],
        "authentication": [],
        "assertionMethod": [],
        "keyAgreement": [],
        "capabilityInvocation": [],
        "capabilityDelegation": [],
        "service": services,
        "metadata": metadata,
    }
    for key in details.get("keys", []):
        if key.get("revoked", False):
            continue
        key_id = _decode_text(key["key_id"])
        material = _normalize_key_material(key)
        controller = _decode_text(key.get("controller") or did.encode("utf-8"))
        vm = {
            "id": key_id,
            "controller": controller,
        }
        if material["type"] == "Jwk":
            vm["type"] = "JsonWebKey2020"
            vm["publicKeyJwk"] = _parse_json_bytes(material["public_key_jwk"])
        else:
            vm["type"] = "Multikey"
            vm["publicKeyMultibase"] = to_multikey(
                material["public_key"],
                material.get("multicodec"),
            )
        doc["verificationMethod"].append(vm)
        for role in key.get("roles", []):
            field = role_map.get(role)
            if field:
                doc[field].append(key_id)
    return doc


def resolve_did(substrate, did: str) -> dict:
    try:
        response = substrate.rpc_request("did_getByString", [did])
    except SubstrateRequestException as exc:
        try:
            result = _query_did_records(substrate, did)
            if result:
                resolved = _success(did, result)
                resolved["didResolutionMetadata"]["warning"] = (
                    "did_getByString failed; resolved through Did.DidRecords storage query"
                )
                return resolved
        except Exception:
            pass
        return {
            "didDocument": None,
            "didDocumentMetadata": {
                "deactivated": False,
                "versionId": 0,
            },
            "didResolutionMetadata": {
                "contentType": None,
                "error": "runtimeApiDecodeError",
                "message": str(exc),
            },
        }

    if isinstance(response, dict):
        result = response.get("result")
        if result:
            return _success(did, result)

    return _not_found()
